sklearn/utils/stats.py

```diff
@@ -1,11 +1,234 @@
+
+            error_higher = np.array([
+                _compute_abs_error(
+                    p_measured=adjusted_percentile,
+                    p_desired=percentile,
+                    idx=percentile_higher_idx,
+                    col=col,
+                )
+                for col in range(n_cols)
+            ])
+            error_lower = np.array([
+                _compute_abs_error(
+                    p_measured=adjusted_percentile,
+                    p_desired=percentile,
+                    idx=percentile_lower_idx,
+                    col=col,
+                )
+                for col in range(n_cols)
+            ])
+
+            percentile_idx = np.where(
+                error_higher >= error_lower,
+                percentile_lower_idx,
+                percentile_higher_idx
+            )
+        elif interpolation == "lower":
+            percentile_idx = percentile_lower_idx
+    else:
+        percentile_idx = percentile_higher_idx
+
+    if interpolation in ("weighted", "linear"):
+        percentile_value_lower = array[
+            sorted_idx[percentile_lower_idx, np.arange(n_cols)],
+            np.arange(n_cols)
+        ]
+        percentile_value_higher = array[
+            sorted_idx[percentile_higher_idx, np.arange(n_cols)],
+            np.arange(n_cols)
+        ]
+
+        if interpolation == "weighted":
+            weight_higher = sorted_weights[
+                percentile_higher_idx, np.arange(n_cols)
+            ]
+            weight_lower = sorted_weights[
+                percentile_lower_idx, np.arange(n_cols)
+            ]
+            weight_normalizer = weight_higher + weight_lower
+            with np.errstate(invalid="ignore"):
+                weight_higher = weight_higher / weight_normalizer
+                weight_higher = _fill_nan(weight_higher, 0)
+                weight_lower = weight_lower / weight_normalizer
+                weight_lower = _fill_nan(weight_lower, 0)
+
+        else:
+            # linear interpolation does not use weights
+            weight_lower = np.zeros(shape=(n_cols,))
+            weight_higher = np.zeros(shape=(n_cols,))
+
+        def _weighted_interpolation(
+            x, x_low, x_high, y_low, y_high, w_low, w_high,
+        ):
+            """Weighted interpolation between 2 values.
+
+            If the weights are null, then it is equivalent to a linear
+            interpolation.
+            """
+            with np.errstate(invalid="ignore"):
+                percentile_value = (
+                    y_low + ((1 - w_low) * (x - x_low)) * (
+                        (y_high - y_low) / (
+                            (1 - w_high) * (x_high - x) +
+                            (1 - w_low) * (x - x_low)
+                        )
+                    )
+                )
+            if np.isnan(percentile_value):
+                # case that percentile is 100
+                return y_high
+            return percentile_value
```

Comment:
Somehow this weighted interpolation will do as on the picture below

![Figure_1](https://user-images.githubusercontent.com/7454015/86778002-26df9d00-c05a-11ea-965e-1371573d6eba.png)

While this closer to the intuition that one would have. I have no idea if this is mathematically correct.
Anybody with some background in interpolation? @jnothman @NicolasHug @thomasjpfan @jeremiedbb @lucyleeow ?